ChemistryLib/PhreeqcInterface.cpp

```diff
@@ -0,0 +1,296 @@
+/**
+ * \copyright
+ * Copyright (c) 2012-2019, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ */
+
+#include <math.h>
+#include <boost/tokenizer.hpp>
+
+#include "BaseLib/ConfigTreeUtil.h"
+#include "Output.h"
+#include "PhreeqcInterface.h"
+#include "ThirdParty/iphreeqc/src/IPhreeqc.h"
+
+namespace ChemistryLib
+{
+void phreeqcIO::doWaterChemistryCalculation(
+    std::vector<GlobalVector*>& process_solutions,
+    ProcessLib::Process const& process, double const dt)
+{
+    setOrUpdateProcessSolutions(process_solutions, process,
+                                         Status::SettingSolutions);
+    setTimeStep(dt);
+
+    writeInputsToFile();
+
+    execute();
+
+    readOutputsFromFile();
+
+    setOrUpdateProcessSolutions(process_solutions, process,
+                                         Status::UpdatingProcessSolutions);
+}
+
+void phreeqcIO::setOrUpdateProcessSolutions(
+    std::vector<GlobalVector*> const& process_solutions,
+    ProcessLib::Process const& process, Status const status)
+{
+    int const nNodes = _solutions.size();
+    for (int node_id = 0; node_id < nNodes; ++node_id)
+    {
+        int process_id = 0;
+        int const variable_id = 0;
+        auto& solution = _solutions[node_id];
+        auto& components = solution._components;
+        for (auto process_solution_iter = process_solutions.begin();
+             process_solution_iter != process_solutions.end();
+             ++process_solution_iter)
+        {
+            auto const& per_process_variables =
+                process.getProcessVariables(process_id);
+            auto const& pv_name =
+                per_process_variables[variable_id].get().getName();
+
+            auto component = std::find_if(components.begin(), components.end(),
+                                          [&pv_name](Component const& c) {
+                                              return c.getName() == pv_name;
+                                          });
+            if (component != components.end())
+            {
+                switch (status)
+                {
+                    case Status::SettingSolutions:
+                        // Set component concentrations.
+                        component->setAmount(
+                            (*process_solution_iter)->get(node_id));
```

Comment:
here you would write
```c++
component.amount = (*process_solution_iter)->get(node_id));
```
No set function is required.

Reply:
Changed.

Reply:
What about the function names `get(node_id)` and `set(node_id, amount)`?

Reply:
The member functions `double EigenVector::get(IndexType rowId) const` and `void EigenVector::set(IndexType rowId, double v)` are invoked here. Do you mean to add another member function in class EigenVector for this special purpose?

Reply:
I see, my bad. I didn't see from the name, that `process_solution_iter` points to a `GlobalVector` pointer.